### ISSUE TITLE
TypedPropertyFromAssignsRector: wrongly (?) assigns `null` to properties

### ISSUE THREAD

**Issue by Aerendir** [rectorphp/rector#7121]
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | 0.12.21 |

## Minimal PHP Code Causing Issue

https://getrector.org/demo/9a9d36e4-7e11-4da2-9ac1-c6128521e2be

## Expected Behaviour

Rector changes the code this way:

```diff
-    /** @var string The number part of the VAT number */
-    private $number;
+    private ?string $number = null;
```

But I expect it changes it this other way:

```diff
-    /** @var string The number part of the VAT number */
-    private $number;
+    private string $number;
```

Am I missing something?

I know the demo is really complex as I had to put in a lot of classes and interfaces.

For a cleaner view of what's happening, see the PR to update Rector (the one that starts to cause the issue) here:
https://github.com/Aerendir/component-value-objects/pull/378

**Comment by TomasVotruba** [comment]
Hi, as in another issue, the code snippet is too long. Please narrow it down to minimal code causing the issue you have.
Should be 1 property, 1 class or interface withing ~30 lines.

**Comment by samsonasik** [comment]
The `$number` doesn't has default value, and never filled by `__construct`, without make it nullable, it will make possible error unitialized value, so the nullable type is on purpose.

**Comment by Aerendir** [comment]
@TomasVotruba  I've updated the snippet to be as thin as possible.
@samsonasik , yes, I understand which is the problem. But the problem is that I don't want this safety: this is something that the developer should decide, not an automatic software.

The purpose of the rule should be to only typehint the property IF POSSIBLE. If it is not possible (because the code cannot be analyzed or because of any other problem, then the rule should do nothing).

The properties are not null because **I want the error to happen**: if the error doesn't happne, I will never know I have an issue in the code.

Adding some sort of failback (as Rector does adding the `null`) is not desirable in this situation (and in many others I'm facing with upgrading Rector on other code bases.

**Comment by samsonasik** [comment]
@Aerendir then it should be skipped, could you create failing fixture with dynamic property filled via trait? Thank you.

**Comment by samsonasik** [comment]
On non-dynamic property set, it should be nullable without __construct, as is, there are a lot of usage that make not nullable will make issue. Skipping should only if there is property set with dynamic property assignment.

**Comment by Aerendir** [comment]
> @Aerendir then it should be skipped, could you create failing fixture PR with dynamic property filled via trait? Thank you.

Mmm... I'm not sure the problem here is the dynamic assignment: if you check the new demo I linked, it has simply the setters that accept `string` while Rector thinks they should also be nullable. I've removed the getters, but also if they have only the `string` typhint (without be nullable), Rector anyway changes the properties and this is, again, wrong.

The problem here is this:

1. I have a class with some properties that have to be valorized;
2. If I call a getter before having called a setter, then I want the error to happen: it tells me that there is something wrong in the way I've implemented the object.
For sure there are situations where it is desirable to be able to call the property before it is valorized, but there are also situations where it is not.

Rctor should not decide on this: this is a decision that has to be taken into account by the developer.

**Comment by samsonasik** [comment]
Then that's the edge case for the specific use case, you can skip it via `$rectorConfig->skip()` method for that :). Make nullable is always a way to ensure there is no destruction in existing code, in many rules rector have. I am closing it as that's is on purpose.

**Comment by Aerendir** [comment]
@samsonasik it is not a specific use case!

It is the way PHP works: there are situation where you want a fallback, and situations where you want it.

Do you always allow `null` returns from you methods?

Please, @TomasVotruba , come here...

Please, think at it: this is not an edge case: you are making a terrible mistake.

**Comment by Aerendir** [comment]
> Make nullable is always a way to ensure there is no destruction in existing code

Sorry, but this is YOUR way of developing: other may prefer to get the errors an be sure each property has EXACTLY the value they want they have, whothout allowing them to have a null value.

If it would be as you describe it, then all methods should be allowed to return null values, and this is not the case.

**Comment by samsonasik** [comment]
There are many people/frameworks that uses rector to upgrade the code, and sooo many crash when no nullable applied on non filled by __construct usage. 

I seen a lot of legacy code for it. 

Dynamic filled should be skipped, I agree, but on exact name, filled in NON `__construct()` or `setUp()`, it should be nullable.

Rector tried to acomodate many use cases as possible, and still there is always a very specific use case of user that wan't different thing, that's user can choose to use the specific rule on specific file or not.

**Comment by Aerendir** [comment]
I understand your point and I understand what you are saying...

But, honestly, the solution, in your opinion, is to do the hard work and be sure a property is called only after it is initialized or allowing a `null` value and let the code "run anyway", also if it is write in the wrong way?

I think that this way of fixing things is an antipattern because let the developer be less cautious when writes the code.

I think that if a property is not initialized and called anyway, the error MUST fail because there is something done wrong by the developer.

The rule is fantastic because it allows to refactor the code and add typehints very fastly, but this behavior is really dangerous because entice developers to be less responsible of the code they write.

Anyway, I'm going to apply the rule to many other code bases and see how it behaves: maybe I'm wrong and this is really an edge case (or only a way of seeing the development, a "philosophy").

For the moment, thank you for the time you dedicated to me and my issue.

**Comment by TomasVotruba** [comment]
The rule behaves this way on purpose. It can only see what is in the code and what can be infered from that.
If you're needs are different on any purpose, use the `skip()` parameter. You can also copy this rule and adjust to your custom one, that meets your specific needs.

**Comment by Aerendir** [comment]
Ok, all clear.

Maybe this behavior can be made configurable... If you like... Or split the rule in two: one to add typehints and one to make properties nullable... I don't know... Anyway, this is your thought and I respect it, also if not fully agree with it.
